modules/xfeatures2d/src/freak.cpp

```diff
@@ -287,8 +287,8 @@ void FREAK_Impl::buildPattern()
         const float dx = patternLookup[orientationPairs[m].i].x-patternLookup[orientationPairs[m].j].x;
         const float dy = patternLookup[orientationPairs[m].i].y-patternLookup[orientationPairs[m].j].y;
         const float norm_sq = (dx*dx+dy*dy);
-        orientationPairs[m].weight_dx = int((dx/(norm_sq))*4096.0+0.5);
-        orientationPairs[m].weight_dy = int((dy/(norm_sq))*4096.0+0.5);
+        orientationPairs[m].weight_dx = int((dx/(norm_sq))*4096.0+0.5*((dx>0)-(dx<0)));
+        orientationPairs[m].weight_dy = int((dy/(norm_sq))*4096.0+0.5*((dy>0)-(dy<0)));
```

Comment:
Please use `cvRound()` instead of 0.5 hack and "int" cast.

Another question is it allowed to weight_dx/dy become zero or negative?
Usually "weight" should not be negative. May be there is lost `abs()` call?

Reply:
weight_dx/dy can be negative to preserve the sense of direction
`direction0 += delta*(orientationPairs[m].weight_dx)/2048;`
`direction1 += delta*(orientationPairs[m].weight_dy)/2048;`
where delta is the difference in mean intensity between two pairs.

Reply:
But after using `cvRound()` tests are getting failed due to bad accuracy.